### ISSUE TITLE
replication-offset-checkpoint.tmp (No such file or directory)

### ISSUE THREAD

**Issue by MichalOrlowski** [spring-projects/spring-kafka#194]
We are using Embedded Kafka for Integration testing. 
What we're getting is `FileNotFoundException` (sometimes) on Jenkins. 
Question - is there a possibility to turn off `ReplicaManager` or so? Maybe there's another solution.

```
2016-10-13 07:53:35.161 ERROR 9469 --- [fka-scheduler-1] kafka.server.ReplicaManager              : [Replica Manager on Broker 0]: Error writing to highwatermark file: 

java.io.FileNotFoundException: /tmp/kafka-3266968904825284552/replication-offset-checkpoint.tmp (No such file or directory)
    at java.io.FileOutputStream.open0(Native Method)
    at java.io.FileOutputStream.open(FileOutputStream.java:270)
    at java.io.FileOutputStream.<init>(FileOutputStream.java:213)
    at java.io.FileOutputStream.<init>(FileOutputStream.java:162)
    at kafka.server.OffsetCheckpoint.write(OffsetCheckpoint.scala:37)
    at kafka.server.ReplicaManager$$anonfun$checkpointHighWatermarks$2.apply(ReplicaManager.scala:874)
    at kafka.server.ReplicaManager$$anonfun$checkpointHighWatermarks$2.apply(ReplicaManager.scala:871)
    at scala.collection.TraversableLike$WithFilter$$anonfun$foreach$1.apply(TraversableLike.scala:778)
    at scala.collection.immutable.Map$Map1.foreach(Map.scala:116)
    at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:777)
    at kafka.server.ReplicaManager.checkpointHighWatermarks(ReplicaManager.scala:871)
    at kafka.server.ReplicaManager$$anonfun$1.apply$mcV$sp(ReplicaManager.scala:153)
    at kafka.utils.KafkaScheduler$$anonfun$1.apply$mcV$sp(KafkaScheduler.scala:110)
    at kafka.utils.CoreUtils$$anon$1.run(CoreUtils.scala:60)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
```

**Comment by artembilan** [comment]
Doesn't that mean that you have a low `ulimit` on that server?
https://easyengine.io/tutorials/linux/increase-open-files-limit/

**Comment by MichalOrlowski** [comment]
No, that's not it. I tried to increase it, but still got same problem.

**Comment by artembilan** [comment]
Well, share, please, as much info then as possible: test-case to play from our side, how much concurrent tests you have, aren't there some process on that server which clean the `/tmp` dir, which Kafka version do you use, etc. etc.. 

Looks like it is more Apache Kafka issue than Spring Kafka...
I'm afraid that we can't switch off `ReplicaManager`...

**Comment by MichalOrlowski** [comment]
I've noticed that this issue occures when Jenkins builds two microservices simultaneously (and both are using Kafka Test).

Question - is there a way to customize /tmp dir name? We need to have possibility building services simultaneously.

**Comment by artembilan** [comment]
As I said before: it is the question directly to Apache Kafka.
```
public EmbeddedZookeeper() {
        this.snapshotDir = kafka.utils.TestUtils..MODULE$.tempDir();
        this.logDir = kafka.utils.TestUtils..MODULE$.tempDir();
```

I'm not familiar with Scala, so can't debug and interpret what they do there...

**Comment by garyrussell** [comment]
It seems unlikely to be related to concurrent builds, in my experience, the `/kafka-n/` directory gets a different `n` on each run.

**Comment by MichalOrlowski** [comment]
Isn't ReplicaManager.checkpointHighWatermarks() method clearing all available replicas?

I think there's possibility to remove /tmp/kafka-n by other thread.

@see core/src/main/scala/kafka/server/ReplicaManager.scala

**Comment by artembilan** [comment]
I still don't understand why do you ask that question to us, Spring community when you see the problem in the Apache Kafka directly? 
Let's move this discussion to StackOverflow: there is nothing from the Spring perspective to fix or improve.

If you insist it is `ReplicaManager`, so please go to the Apache Kafka community and ask there.
That is so low level of the integration that we are just not aware of.

Sorry, but it looks like we (at least me) are useless for your on the topic and I don't understand why do you spend time with us not Apache Kafka community?

Would be glad to see some cross-link from there to widen knowledge in this Kafka topic.

Thanks

**Comment by MichalOrlowski** [comment]
Right now I just mocked ReplicaManager.shutdown() method to return false.

Thanks for you comments.

**Comment by hoaz** [comment]
I have found this thread via google looking for solution. Did not find it here, but this is the only place which refers to problem I encountered.
I did analysis and want to share it with others. Here is what happens:
1. kafka TestUtils.tempDirectory method is used to create temporary directory for embedded kafka broker. It also registers shutdown hook which deletes this directory when JVM exits.
2. when unit test finishes execution it calls System.exit, which in turn executes all registered shutdown hooks

If kafka broker is still running it will attempt to write/read data in a dir which is deleted and produce different FileNotFound exceptions.

**Solution**: shutdown embedded kafka broker at the end of the test before System.exit is called.

If KafkaEmbedded rule is used properly it will call KafkaEmbedded#after method, which destroys broker before System#exit is called.

I use KafkaEmbedded class in Spring integration test and create it as bean. Unfortunately spring context is destroyed in shutdown hook as well and it happens concurrently with other shutdown hooks, so kafka log directory is destroyed before embedded kafka broker is down. I did not find proper solution yet for this usage scenario.

**Comment by artembilan** [comment]
@hoaz ,

thank you for your report.
The question is: why do you use `KafkaEmbedded` as a bean?
Why `@ClassRule` is not sufficient for you?
```
@ClassRule
public static KafkaEmbedded embeddedKafka = new KafkaEmbedded(1, true, INT_KEY_TOPIC, STRING_KEY_TOPIC);
```

> If kafka broker runs at the end of unit test it will attempt to write/read data in a dir which is deleted and produces different FileNotFound exceptions.

Well, it would be better to stop writing in the end of test, e.g. call `stop()` of the endpoints or verify that sent message has been reached broker before existing from the test. Right?

**Comment by hoaz** [comment]
Because I have multiple integration tests which share the same Spring test context. I do not want to bring up and shut down kafka broker in each test but rather delegate it to Spring, which caches spring context between tests.
This approach speeds up integration tests with common test Spring context significantly.

**Comment by artembilan** [comment]
Would you mind to share the code how do you do that?
According `@ClassRule` JavaDocs we have:
```
* For example, here is a test suite that connects to a server once before
 * all the test classes run, and disconnects after they are finished:
 * <pre>
 * &#064;RunWith(Suite.class)
 * &#064;SuiteClasses({A.class, B.class, C.class})
 * public class UsesExternalResource {
 *     public static Server myServer= new Server();
 *
 *     &#064;ClassRule
 *     public static ExternalResource resource= new ExternalResource() {
```
So, placing the `KafkaEmbedded` rule there on a suite level, should make us happy. Isn't it?

**Comment by hoaz** [comment]
OK, this is a valid solution, there is only one thing I do not like about it.
Our build configuration uses default surefire configuration which scans and runs all integration tests that match specific wildcard. So every time new test is added developer needs to make sure it is included in test suite, as well as changes to build configuration should be made.

It would be really good if spring-kafka-test could support embedded kafka as a bean, especially taking into account project name :)

Here is how we run embedded kafka broker inside of spring container:

```
@Configuration
public class EmbeddedKafkaConfiguration {

    @Bean(destroyMethod = "after")
    public static KafkaEmbedded kafkaEmbedded() throws Exception {
        return new KafkaEmbeddedBean(1, true, 1, TestApplication.TOPIC_NAME);
    }

    private static final class KafkaEmbeddedBean extends KafkaEmbedded {

        public KafkaEmbeddedBean(int count, boolean controlledShutdown, int partitions, String... topics) throws Exception {
            super(count, controlledShutdown, partitions, topics);
            before();
            System.setProperty("kafka.broker.list", getBrokersAsString());
        }
        
    }

}

```

**Comment by artembilan** [comment]
Good, thank you for sharing that and for your ideas!

So, doesn't it mean that that `destroyMethod = "after"` may be called already after that:
```
Runtime.getRuntime().addShutdownHook(new Thread() {
           public void run() {
               Utils.delete(file);
           }
});
```
And what we get there is only some ERROR in logs?

**Comment by hoaz** [comment]
Yes, Spring context is being closed in parallel with other shutdown hooks, they are executed in separate threads (see java.lang.ApplicationShutdownHooks#runHooks). Spring shutdown hook is rather slow, so in most cases `destroyMethod = "after"` will be called after directories are already gone.

And yes again, you are right, ERRORs in logs do not fail build, they just sitting there making me nervous each time I analyze logs.

**Comment by artembilan** [comment]
OK. Great!
So, now, please, share the latest ERROR for that case.
I think I have an idea what to do in the `KafkaEmbedded.after()` to avoid such a late shutdown because it just does not make sense any more.

**Comment by hoaz** [comment]
I have two errors actually.

This one originates from `KafkaEmbedded.after()`:

> 15:17:24,975 [                      Thread-8] FATAL ReplicaManager:118 - [Replica Manager on Broker 0]: Error writing to highwatermark file: 
> java.io.FileNotFoundException: /tmp/kafka-1318430730057043027/replication-offset-checkpoint.tmp (No such file or directory)
> 	at java.io.FileOutputStream.open(Native Method)
> 	at java.io.FileOutputStream.<init>(FileOutputStream.java:221)
> 	at java.io.FileOutputStream.<init>(FileOutputStream.java:171)
> 	at kafka.server.OffsetCheckpoint.write(OffsetCheckpoint.scala:49)
> 	at kafka.server.ReplicaManager$$anonfun$checkpointHighWatermarks$2.apply(ReplicaManager.scala:948)
> 	at kafka.server.ReplicaManager$$anonfun$checkpointHighWatermarks$2.apply(ReplicaManager.scala:945)
> 	at scala.collection.TraversableLike$WithFilter$$anonfun$foreach$1.apply(TraversableLike.scala:733)
> 	at scala.collection.immutable.Map$Map1.foreach(Map.scala:116)
> 	at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:732)
> 	at kafka.server.ReplicaManager.checkpointHighWatermarks(ReplicaManager.scala:945)
> 	at kafka.server.ReplicaManager.shutdown(ReplicaManager.scala:964)
> 	at kafka.server.KafkaServer$$anonfun$shutdown$7.apply$mcV$sp(KafkaServer.scala:590)
> 	at kafka.utils.CoreUtils$.swallow(CoreUtils.scala:78)
> 	at kafka.utils.Logging$class.swallowWarn(Logging.scala:94)
> 	at kafka.utils.CoreUtils$.swallowWarn(CoreUtils.scala:48)
> 	at kafka.utils.Logging$class.swallow(Logging.scala:96)
> 	at kafka.utils.CoreUtils$.swallow(CoreUtils.scala:48)
> 	at kafka.server.KafkaServer.shutdown(KafkaServer.scala:590)
> 	at org.springframework.kafka.test.rule.KafkaEmbedded.after(KafkaEmbedded.java:173)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.lang.reflect.Method.invoke(Method.java:606)
> 	at org.springframework.beans.factory.support.DisposableBeanAdapter.invokeCustomDestroyMethod(DisposableBeanAdapter.java:300)
> 	at org.springframework.beans.factory.support.DisposableBeanAdapter.destroy(DisposableBeanAdapter.java:226)
> 	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.destroyBean(DefaultSingletonBeanRegistry.java:499)
> 	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.destroySingleton(DefaultSingletonBeanRegistry.java:475)
> 	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.destroySingletons(DefaultSingletonBeanRegistry.java:443)
> 	at org.springframework.context.support.AbstractApplicationContext.destroyBeans(AbstractApplicationContext.java:1078)
> 	at org.springframework.context.support.AbstractApplicationContext.doClose(AbstractApplicationContext.java:1052)
> 	at org.springframework.context.support.AbstractApplicationContext$1.run(AbstractApplicationContext.java:970)

And this one from Kafka LogCleaner, I guess it will be hard to get rid of it:

> 15:17:24,828 [    kafka-log-cleaner-thread-0] ERROR LogCleaner:105 - [kafka-log-cleaner-thread-0], Error due to 
> java.io.FileNotFoundException: /tmp/kafka-1318430730057043027/cleaner-offset-checkpoint (No such file or directory)
> 	at java.io.FileInputStream.open(Native Method)
> 	at java.io.FileInputStream.<init>(FileInputStream.java:146)
> 	at java.io.FileReader.<init>(FileReader.java:72)
> 	at kafka.server.OffsetCheckpoint.read(OffsetCheckpoint.scala:86)
> 	at kafka.log.LogCleanerManager$$anonfun$allCleanerCheckpoints$1.apply(LogCleanerManager.scala:81)
> 	at kafka.log.LogCleanerManager$$anonfun$allCleanerCheckpoints$1.apply(LogCleanerManager.scala:81)
> 	at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:241)
> 	at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:241)
> 	at scala.collection.Iterator$class.foreach(Iterator.scala:893)
> 	at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
> 	at scala.collection.MapLike$DefaultValuesIterable.foreach(MapLike.scala:206)
> 	at scala.collection.TraversableLike$class.flatMap(TraversableLike.scala:241)
> 	at scala.collection.AbstractTraversable.flatMap(Traversable.scala:104)
> 	at kafka.log.LogCleanerManager.allCleanerCheckpoints(LogCleanerManager.scala:81)
> 	at kafka.log.LogCleanerManager$$anonfun$grabFilthiestCompactedLog$1.apply(LogCleanerManager.scala:92)
> 	at kafka.log.LogCleanerManager$$anonfun$grabFilthiestCompactedLog$1.apply(LogCleanerManager.scala:89)
> 	at kafka.utils.CoreUtils$.inLock(CoreUtils.scala:213)
> 	at kafka.log.LogCleanerManager.grabFilthiestCompactedLog(LogCleanerManager.scala:89)
> 	at kafka.log.LogCleaner$CleanerThread.cleanOrSleep(LogCleaner.scala:234)
> 	at kafka.log.LogCleaner$CleanerThread.doWork(LogCleaner.scala:220)
> 	at kafka.utils.ShutdownableThread.run(ShutdownableThread.scala:63)

**Comment by artembilan** [comment]
@hoaz ,

you don't need
```
System.setProperty("kafka.broker.list", getBrokersAsString());
```
The `KafkaEmbedded` does that like:
```
System.setProperty(SPRING_EMBEDDED_KAFKA_BROKERS, getBrokersAsString());
```
in the end of `before()`.

**Comment by hoaz** [comment]
Right, long story short, test `@Configuration` I have provided above is appended to main application configuration. And this line overrides non-test property kafka.broker.list, which we use in DEV, UAT, PROD.

For test only purposes SPRING_EMBEDDED_KAFKA_BROKERS is sufficient.

**Comment by artembilan** [comment]
@hoaz,

please, take a look to the PR related #294 .

Instead of `destroy()` I perform `after()` from the `stop()`. That is done much earlier than any shutdown hooks.

**Comment by hoaz** [comment]
OK, so this fix add possibility to use `KafkaEmbedded` as a bean, without any additional effort. Which is a nice improvement.

But I am afraid it does not fix original problem. Look at my first stack trace, the very last line is an entry point of shutdown hook:
```
at org.springframework.context.support.AbstractApplicationContext$1.run(AbstractApplicationContext.java:970)
```

Which means `Lifecycle#stop` will be called from `AbstractApplicationContext#doClose`, which is called from shutdown hook. It is already too late, Spring shutdown hook is already running, which means other shutdown hooks either run in parallel or already finished.
I do no think you can fix it in `KafkaEmbedded`, this is how Spring is designed.

**Comment by artembilan** [comment]
Ah! I see:
```
 * @see #close()
 * @see #registerShutdownHook()
 */
protected void doClose() {
...
	// Publish shutdown event.
	publishEvent(new ContextClosedEvent(this));
...
	getLifecycleProcessor().onClose();
...
	destroyBeans();
```
So, for the bean it doesn't matter if we close resource in the `stop()` or `destroy()` methods.

Thank you!


Need to think more...

I wonder if you can provide some test-case to play from our side, because I can't reproduce somehow...

Thank you!

**Comment by artembilan** [comment]
Well, looking to this one more time, I'd say that it is Apache Kafka issue.
They can't make an assumption when we are going to call `kafkaServer.shutdown()`, therefore it doesn't look consistent to hook `/tmp` removal unconditionally.

IMO that should be done in the `kafkaServer.shutdown()` as a consequence of all shutdowned Kafka resources: https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/KafkaServer.scala#L578.

Or, they should just let us do that manually, like it is right now in our `KafkaEmbedded`:
```
CoreUtils.delete(kafkaServer.config().logDirs());
```

For this issue I only consider the polishing for the proposed bean-aware solution and maybe `@EmbeddedKafka` annotation for the test class.

**Comment by hoaz** [comment]
Yeah, I did same analysis and came to conclusion that Kafka devs introduced this problem here: https://issues.apache.org/jira/browse/KAFKA-1258

Most probably they did not think about all scenarios at that time, but this code made it to 2017 and still alive.

`kafkaServer.shutdown()` should not remove folders because kafka in non-test environment should be restartable. But in test environment outside test controller (e.g. JUnit rule) should take care of cleanup.

**Comment by kjkrol** [comment]
I found workaround. I mix both approaches: provided annotation + custom bean shared among all int tests/specs. I know that it look weird, but works for me.

`@Component
@Profile(TEST)
@EmbeddedKafka(partitions = 1, topics = ['asset-change', 'schedule-changed'], controlledShutdown = true)
class EmbeddedKafkaProvider {

    final KafkaEmbedded kafkaEmbedded

    EmbeddedKafkaProvider(@Value('${scheduling.kafka.consumer.topic}') String schedulingTopic,
                          @Value('${asset.kafka.consumer.topic}') String assetTopic) {
        kafkaEmbedded = new KafkaEmbedded(1, true, 1, schedulingTopic, assetTopic)
        kafkaEmbedded.before()
    }

}`

**Comment by jdpgrailsdev** [comment]
Adding an update here in case anyone else stumbles upon this and needs some help.  I am using the `1.2.3.RELEASE` (due to compatibility issues with the version of Kafka I am running) and ran into this issue.  After digging into it, it appears that the issue is the fact that the `KafkaEmbedded` class calls the `shutdown()` method on the underlying Apache `KafkaServer` class when you call `after()` to shut down the embedded server.  Calling `shutdown()` itself is not the issue -- the issue is that the `shutdown()` method calls the default implementation of `shutdown()` on the `ReplicaManager`, which in turn defaults the flag that controls writing the high water mark file to `true`.  You can call `shutdown()` on the `ReplicaManager` with an optional `false` parameter to skip writing the high water mark file (which is really not needed for unit/integration tests).  To get around this issue, I used a little reflection:

```java
final KafkaServer server = embeddedKafka.getKafkaServers().stream().findFirst().orElse(null);
if(server != null) {
    server.replicaManager().shutdown(false);
    final Field replicaManagerField = server.getClass().getDeclaredField("replicaManager");
    if(replicaManagerField != null) {
        replicaManagerField.setAccessible(true);
        replicaManagerField.set(server, null);
     }
}
embeddedKafka.after();
```

The code above explicitly calls `shutdown()` on the `ReplicaManager` used by the `KafkaServer` passing `false` as the parameter and then replaces the `ReplicaManager` instance in the `KafkaServer` with `null` so it does not get called again with the `true` parameter value when `after()` is called on the `embeddedKafka` instance.  This is totally a hack and definitely has more to do with the API exposed by Apache Kafka than the Spring library.  Hopefully, this is easier to fix in more recent versions of both this library and Apache Kafka.  However, if you are stuck on this version (or a similar one) like me, the solution above should solve the error outlined in this GitHub issue.

**Comment by artembilan** [comment]
Yeah... Would be better to raise such an issue in the Apache Kafka per se.

Great finding @jdpgrailsdev though! Thank you for sharing!

**Comment by hstaos** [comment]
I found KafkaScheduler.class in the stack traces, so I adjust the Kafka server property replica.high.watermark.checkpoint.interval.ms to a very large number (e.g. 100000000).  

This work for me.

**Comment by garyrussell** [comment]
@hstaos Thanks for sharing.

**Comment by senthilsivanath** [comment]
@jdpgrailsdev and @hstaos 

Guys, you saved my day.

**Comment by jdpgrailsdev** [comment]
Based on what @hstaos found, it seems like we could patch `KafkaEmbedded` to explicitly set `replica.high.watermark.checkpoint.interval.ms` in the `brokerConfigProperties`.  Unfortunately, in the version of `spring-kafka-test` that I am using (`1.2.3.RELEASE`) there is no way to modify the `brokerConfigProperties` passed to the `TestUtils.createServer()` method, as it is all encapsulated in the `before()` method of the `KafkaEmbedded` class 😢 .  If we feel it is worth while to patch fix an older release, I can certainly put a PR together to add that server property.  Otherwise, it does not appear that those properties can be set in the older releases of this library.

**Comment by artembilan** [comment]
@jdpgrailsdev ,

Yes, you really feel free to raise a PR, but at the same time consider to upgrade at least to the `1.3.x` line for better threading model and more improvements there. The newer Kafka Client can talk to older Brokers. That's the point for newer Spring Kafka version to rely on.

**Comment by jdpgrailsdev** [comment]
@artembilan I would love to be able to upgrade to a newer version of Spring Kafka, but unfortunately I need one that works with the 0.10.2.1 version of Kafka.  Therefore, if I am reading the compatibility chart on the project page for Spring Kafka, I am stuck at the 1.2.3.RELEASE version based on my Kafka version.  Not sure if you can cut a point release of 1.2.3.RELEASE.  If you can, I can certainly submit a PR.  If not, no worries.

**Comment by artembilan** [comment]
No, that table is for the Spring Kafka <-> Kafka Client compatibility.
As long as you are going to use Spring Kafka `1.3.x`, you definitely get a newer client, but it still can talk to an old Broker. That's the point.

**Comment by garyrussell** [comment]
I will add a note to the matrix explaining that all current versions can talk to brokers newer than 0.9.

**Comment by garyrussell** [comment]
Heres [the link](https://cwiki.apache.org/confluence/display/KAFKA/Compatibility+Matrix).

For broker 0.10.2 it says:

Broker 0.10.2
- Basic client compatibility:
  - Java: any version

**Comment by NeveHanter** [comment]
Sorry for digging out that issue but I think this problem wasn't properly resolved.

Does anyone of people involved in this issue reported the problem with the suggestion for file cleanup during unit testing on Kafka's bug tracker?

**Comment by gmandnepr** [comment]
I have faced the same issue with spring-kafka-test 2.1.9.RELEASE

During troubleshooting I have found that org.springframework.kafka.test.rule.KafkaEmbedded#after seems to throw Throwable (or somehow calls System.exit) from kafkaServer.shutdown() before line CoreUtils.delete(kafkaServer.config().logDirs()) that is needed to fix the exception.

I have implemented custom TestExecutionListener
`public class CustomEmbeddedKafkaTestExecutionListener implements TestExecutionListener {

    private KafkaEmbedded kafkaEmbedded;

    @Override
    public void beforeTestClass(TestContext testContext) throws Exception {

        kafkaEmbedded = new KafkaEmbedded(1, true, 1, "myTopic");
        kafkaEmbedded.before();
    }

    @Override
    public void afterTestClass(TestContext testContext) throws Exception {

        kafkaEmbedded.after();

    }
}`
That does exactly the same, but for some reason after does not throw anything and my test is not failing any more.

**Comment by artembilan** [comment]
@gmandnepr ,

Your code really doesn't bring any difference because you still call `kafkaEmbedded.after();` as is. And that one performs `kafkaServer.shutdown();` and only then  `CoreUtils.delete(kafkaServer.config().logDirs())`.

Doesn't this sound that we should call that `CoreUtils.delete(kafkaServer.config().logDirs())` in the `KafkaEmbedded.after()` before the `kafkaServer.shutdown();`?

Thanks.

**Comment by CluEleSsUK** [comment]
I had a similar issue to this, but was using Cucumber, so the failure was actually causing a build failure (despite all the tests passing).

The solution I had was to add a shutdown hook before Spring shuts down as below:

``@Component
public class EmbeddedKafkaLifecycle {

  @Autowired
  KafkaEmbedded kafkaEmbedded;

  @PreDestroy
  public void onDestroy() {
    kafkaEmbedded.after();
  }

}

**Comment by gmandnepr** [comment]
I am currently having exactly the same. All tests passes (Junit5), but test phase returns code 1 and this fails the build.
Could it be something in kafka calls System.exit(1)?
I have replaced kafkaEmbedded.after(); with my custom code similar to after(), but catches Throwable and logs that, but non of catch blocks was executed

**Comment by CluEleSsUK** [comment]
Yeah, there is a fatal call in the `LogManager`, but this is in kafka itself (not spring-kafka)

```
  def handleLogDirFailure(dir: String) {
    info(s"Stopping serving logs in dir $dir")
    logCreationOrDeletionLock synchronized {
      _liveLogDirs.remove(new File(dir))
      if (_liveLogDirs.isEmpty) {
        fatal(s"Shutdown broker because all log dirs in ${logDirs.mkString(", ")} have failed")
        Exit.halt(1)
      }
```

Could this be a timing issue with Spring deleting the logs while the brokers are still up?

**Comment by nightswimmings** [comment]
I have been looking the ticket for the error in Kafka jiras but I can find it. Does anybody knows whre to locate it in order to track the progress in the resolution of this issue?

**Comment by jvwilge** [comment]
We use JUnit 5 with Karate on our projects and we're using the `@EmbeddedKafka` annotation and ran into the same problem as the issue starter. Adding a `@DirtiesContext` on a test method eventually did the trick and the error disappeared.

**Comment by josecv** [comment]
This issue is still happening as of `spring-kafka` `2.2.5.RELEASE`

We did some debugging on it, and it seems to come from a conflicting set of JVM shutdown hooks.

`spring-kafka-test` is using [TestUtils](https://github.com/spring-projects/spring-kafka/blob/a7a4df7c30da4bc24ba9c1a1dd91a7ad3914f7ef/spring-kafka-test/src/main/java/org/springframework/kafka/test/EmbeddedKafkaBroker.java#L239) to set up the kafka properties. Under the hood, this is going to `org.apache.kafka.test.TestUtils.tempDirectory` which registers a [shutdown hook](https://github.com/apache/kafka/blob/bebcbe3a049f78c4184404f2dfb8b4150233856e/clients/src/test/java/org/apache/kafka/test/TestUtils.java#L261) that is meant to clean up the temp directory upon JVM shutdown. If this hooks runs with the broker still up, the LogManager does a `System.exit` [as described](https://github.com/spring-projects/spring-kafka/issues/194#issuecomment-419023072).

`spring-kafka-test` is [using the context lifecycle](https://github.com/spring-projects/spring-kafka/blob/a7a4df7c30da4bc24ba9c1a1dd91a7ad3914f7ef/spring-kafka-test/src/main/java/org/springframework/kafka/test/context/EmbeddedKafkaContextCustomizer.java#L111) to [shutdown the broker](https://github.com/spring-projects/spring-kafka/blob/a7a4df7c30da4bc24ba9c1a1dd91a7ad3914f7ef/spring-kafka-test/src/main/java/org/springframework/kafka/test/EmbeddedKafkaBroker.java#L321) — when `@DirtiesContext` is set, the context is destroyed before the JVM starts shutting down. That way, once the `TestUtils` shutdown hook runs, the broker is actually down and everything works.

Unfortunately, when `DirtiesContext` is not set, Spring’s context teardown also relies on a JVM shutdown hook.
As the shutdown hook execution order is not guaranteed, this means we have a race condition between Spring shutting down the broker and TestUtils destroying the tempdir.

If the temp directory is cleaned prior to Spring’s context teardown, the Kafka brokers would be left running and the LogManager will eventually be met with an empty directory. Hence the `System.exit`

We worked around this by using a [`SecurityManager`](https://stackoverflow.com/questions/5549720/how-to-prevent-calls-to-system-exit-from-terminating-the-jvm) to forbid `System.exit(1)` in our tests. This ensures that spring has a chance to shutdown the broker, even if the broker wants to die prematurely. Likely a longer term fix would be to have the `EmbeddedKafkaBroker` manage the lifecycle of the temp directory itself, instead of delegating its creation to `TestUtils`

**Comment by artembilan** [comment]
Well, is it even possible to manage that ourselves?
Do you have any clue how could we overcome that Kafka's `TestUtils` limitation?

You are more than welcome to contribute such a feature! (As well as any other though 😉 )

**Comment by josecv** [comment]
Hmm, I'd love to put in that PR, but I have no idea what the fix would be 🙃 

Unfortunately it seems like `TestUtils` [doesn't have](https://github.com/apache/kafka/blob/bebcbe3a049f78c4184404f2dfb8b4150233856e/core/src/test/scala/unit/kafka/utils/TestUtils.scala#L265) the option to use a pre-supplied directory. I suppose you could not use `createBrokerConfig` at all and construct properties yourself. However, it looks like Zookeeper [might have](https://github.com/apache/kafka/blob/bebcbe3a049f78c4184404f2dfb8b4150233856e/core/src/test/scala/unit/kafka/zk/EmbeddedZookeeper.scala#L39) the same problem and it doesn't seem like supplying your own properties is an option there.

**Comment by josecv** [comment]
I definitely think this warrants reopening the issue though, as a few others are also reporting this problem.

**Comment by artembilan** [comment]
Thanks for all your feedback!
Yes, we can reopen it, but it doesn't look like we can have a solution from our side until Apache Kafka provides for us some hooks to configure those temporary directories.

**Comment by alinturbut** [comment]
I think I've found a workaround for the execution fail when tests run in parallel. Each project running must have the spring.application.name defined uniquely.

```
@EmbeddedKafka(brokerProperties = {
    "log.dir=/tmp/kafka-data/${spring.application.name}"
})
```

**Comment by garyrussell** [comment]
Thanks for the tip.

Unfortunately, that won't work when using `@EmebddedKafka` with JUnit5 alone (no Spring context) or with non-Boot Spring tests.

I think we can add a random value in the broker itself.

**Comment by garyrussell** [comment]
Hmmm - not sure why that would be necessary; Kafka already randomizes the name

    /tmp/kafka-3266968904825284552

**Comment by alinturbut** [comment]
I know it does, but it still throws an error sometimes, the tests running in parallel might override the folders, somehow. Haven't invested time investigating into it, but for me this workaround fixed the failing tests.

**Comment by garyrussell** [comment]
Good to know; thanks.

**Comment by Rouche** [comment]
Hi

We are stuck with randomly failing tests. I tryed every solution here without success, Our builds randomly fail with this stack trace. Wich is mentionned by @hoaz above. Its not the same exact stack but i think its the same problem.

```
java.nio.file.NoSuchFileException: C:\Users\rouche\AppData\Local\Temp\kafka-10001186863277274754\cleaner-offset-checkpoint
	at java.base/sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:85)
	at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:103)
	at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:108)
	at java.base/sun.nio.fs.WindowsFileSystemProvider.newByteChannel(WindowsFileSystemProvider.java:235)
	at java.base/java.nio.file.Files.newByteChannel(Files.java:370)
	at java.base/java.nio.file.Files.newByteChannel(Files.java:421)
	at java.base/java.nio.file.spi.FileSystemProvider.newInputStream(FileSystemProvider.java:420)
	at java.base/java.nio.file.Files.newInputStream(Files.java:155)
	at java.base/java.nio.file.Files.newBufferedReader(Files.java:2838)
	at java.base/java.nio.file.Files.newBufferedReader(Files.java:2870)
	at kafka.server.checkpoints.CheckpointFile.liftedTree2$1(CheckpointFile.scala:87)
	at kafka.server.checkpoints.CheckpointFile.read(CheckpointFile.scala:86)
	at kafka.server.checkpoints.OffsetCheckpointFile.read(OffsetCheckpointFile.scala:61)
	at kafka.log.LogCleanerManager.$anonfun$allCleanerCheckpoints$2(LogCleanerManager.scala:135)
	at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:244)
	at scala.collection.Iterator.foreach(Iterator.scala:941)
	at scala.collection.Iterator.foreach$(Iterator.scala:941)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1429)
	at scala.collection.MapLike$DefaultValuesIterable.foreach(MapLike.scala:213)
	at scala.collection.TraversableLike.flatMap(TraversableLike.scala:244)
	at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:241)
	at scala.collection.AbstractTraversable.flatMap(Traversable.scala:108)
	at kafka.log.LogCleanerManager.$anonfun$allCleanerCheckpoints$1(LogCleanerManager.scala:133)
	at kafka.utils.CoreUtils$.inLock(CoreUtils.scala:251)
	at kafka.log.LogCleanerManager.allCleanerCheckpoints(LogCleanerManager.scala:141)
	at kafka.log.LogCleanerManager.$anonfun$grabFilthiestCompactedLog$1(LogCleanerManager.scala:172)
	at kafka.utils.CoreUtils$.inLock(CoreUtils.scala:251)
	at kafka.log.LogCleanerManager.grabFilthiestCompactedLog(LogCleanerManager.scala:169)
	at kafka.log.LogCleaner$CleanerThread.cleanFilthiestLog(LogCleaner.scala:313)
	at kafka.log.LogCleaner$CleanerThread.doWork(LogCleaner.scala:300)
	at kafka.utils.ShutdownableThread.run(ShutdownableThread.scala:82)
```

**Comment by garyrussell** [comment]
>at kafka.utils.ShutdownableThread.run(ShutdownableThread.scala:82)

This is occurring during shutdown; why would that cause a test to fail?

**Comment by Rouche** [comment]
No idea, the exception must happens before the vm exits.

That shutdown is Kafka's not the test itself.

But the test is marked failed and the build too of course.

**Comment by garyrussell** [comment]
Please provide a test case (and detailed environment information) that reliably exhibits this behavior.

**Comment by Rouche** [comment]
I tryed all day to create a minimalist JHipster project where we have this bug on all our services that use Kafka and i am unable to reproduce. This is disturbing. I need to find whats the difference.

**Comment by Rouche** [comment]
Finally reproduced.
https://github.com/Rouche/spring-kafka-bug-194

Its a JHipster generated app.
Little note: if i put `KafkaErrorProducerTest` as an IT test `KafkaErrorProducerTestIT` i cannot reproduce.

The stack trace with `TestUtils` always happens. The one with `ShutdownableThread` is once out of 10 on my system. Windows 10 with an ok machine.

**Comment by chrgue** [comment]
> I think I've found a workaround for the execution fail when tests run in parallel. Each project running must have the spring.application.name defined uniquely.
> 
> ```
> @EmbeddedKafka(brokerProperties = {
>     "log.dir=/tmp/kafka-data/${spring.application.name}"
> })
> ```

@alinturbut This helped us a lot. Thanks!

**Comment by Rouche** [comment]
This create another error:
```
KafkaException: Failed to acquire lock on file .lock in C:\tmp\kafka-data\MyService. A Kafka instance in another process or thread is using this directory.
```

**Comment by mhyeon-lee** [comment]
I solved this problem.

As in https://github.com/spring-projects/spring-kafka/issues/194#issuecomment-501003852, the tmp folder is removed in JVM shutdown, and when using Kafka in another hook, Embedded Kafka calls `Runtime.getRuntime().halt(1)`.

The following code can prevent `halt(1)` from ending the test.

```java
Exit.setHaltProcedure((statusCode, message) -> {
    if (statusCode != 1) {
        Runtime.getRuntime().halt(statusCode);
    }
});
```

**Comment by garyrussell** [comment]
Interesting; thanks.

**Comment by Rouche** [comment]
Where is the Exit object come from?
`kafka.utils`
or
`org.apache.kafka.common.utils`

**Comment by gustavomaia** [comment]
> I solved this problem.
> 
> As in [#194 (comment)](https://github.com/spring-projects/spring-kafka/issues/194#issuecomment-501003852), the tmp folder is removed in JVM shutdown, and when using Kafka in another hook, Embedded Kafka calls `Runtime.getRuntime().halt(1)`.
> 
> The following code can prevent `halt(1)` from ending the test.
> 
> ```java
> Exit.setHaltProcedure((statusCode, message) -> {
>     if (statusCode != 1) {
>         Runtime.getRuntime().halt(statusCode);
>     }
> });
> ```

Hey, I am with the same issue so I am trying to understand your comment, where are you adding this, and how are you using the security manager? 

Thanks in advance.

**Comment by mhyeon-lee** [comment]
@Rouche  
[org.apache.kafka.common.utils.Exit](https://github.com/a0x8o/kafka/blob/b0d7afc28d6d955620b23ae810656d4f7d4215a6/clients/src/main/java/org/apache/kafka/common/utils/Exit.java#L85)

When JVM shutdown Hook is running, kafka log file is deleted and Exit.halt (1) is called when other shutdown hook accesses kafka log file at the same time.

Since halt is called here and status is 1, we only defend against 1.
https://github.com/a0x8o/kafka/blob/master/core/src/main/scala/kafka/log/LogManager.scala#L193

An error log may occur, but the test will not fail because the command is propagated to Runtime.halt.

If you encounter a situation where the test fails with a different status value, you can add defense code.

**Comment by mhyeon-lee** [comment]
@gustavomaia
Since it is a static method setting, you can put it anywhere you want.
However, this setting must be executed before the entire test is completed.

```java
class SomeTest {
    static {
        Exit.setHaltProcedure((statusCode, message) -> {
            if (statusCode != 1) {
                Runtime.getRuntime().halt(statusCode);
            }
        });
    }

    @Test
    void test1()  {
     }

     @Test
     void test2() {
     }
}
```

Even using the SecurityManager, the test can fail.
Exit / halt can be prevented by throwing Exception in `checkExit` of SecurityManager.
However, the exception thrown in JVM shutdownHook cannot be `catch` and the test fails.
